pyramid_openapi3/__init__.py

```diff
@@ -42,48 +39,24 @@ def openapi_view(view: View, info: ViewDeriverInfo) -> t.Optional[View]:
     If `openapi=True` is passed to `@view_config`, this decorator will:
 
     - validate request and submit results into request.openapi_validated
-    - validate response and raise an Exception if errors are found
+    - Only request is validated here. The response is validated inside a tween,
+    so that other tweens can intercept the response, and only the final
+    response is validated against the openapi spec.
     """
     if info.options.get("openapi"):
 
         def wrapper_view(context, request):
             # Validate request and attach all findings for view to introspect
+            request.environ["pyramid_openapi3.validate_response"] = True
             settings = request.registry.settings["pyramid_openapi3"]
             open_request = PyramidOpenAPIRequest(request)
             request.openapi_validated = settings["request_validator"].validate(
                 open_request
             )
-
-            try:
-                if request.openapi_validated.errors and getattr(
-                    request.registry, "_pyramid_openapi3_validation_view_name", None
-                ):
-                    # handle request errors
-                    view_name = (
-                        request.registry._pyramid_openapi3_validation_view_name  # noqa: SF01
-                    )
-                    response = render_view_to_response(
-                        context, request, name=view_name, secure=False
-                    )
-                else:
-                    # Do the view
-                    request.openapi_validated.raise_for_errors()
-                    response = view(context, request)
-
-            except HTTPException as exc:
-                # If view raises one of the HTTPExceptions, catch it
-                # so that we can validate the response and hence make sure
-                # all possible responses are documented in openapi spec.
-                response = exc
-
-            # Validate response and raise if an error is found
-            open_response = PyramidOpenAPIResponse(response)
-            result = settings["response_validator"].validate(
-                request=open_request, response=open_response
-            )
-            result.raise_for_errors()
-
-            return response
+            if request.openapi_validated.errors:
+                raise RequestValidationError(errors=request.openapi_validated.errors)
```

Comment:
If you down the line process exception in Sentry, Stackdriver etc, it's really helpful if you also have request.openapi_validated.body and similar, so I would pass request.openapi_validated and then defer parsing values in the exception handler.

Reply:
As far as I know, in case of a RequestValidationError, `request.openapi_validated.parameters` will only ever have valid parameters and `request.openapi_validated.body` is only populated when the `body` is valid. So any parameter or body content that errors won't be available in on the `ValidationResult`. So all you would get are valid parameters and valid body. 

I was not sure, whether to pass the `ValidationResult` or just the errors to the exception. I couldn't find any useful values on a `ValidationResult` other than errors (everything I would probably interrogate would be available on the request object only), so I opted for that to be on the safe side. My concern was rather about, what if it is a really large body object, and do I really want that to be passed around? (probably not an issue, but I assumed not passing it through is safer :) )

Reply:
> My concern was rather about, what if it is a really large body object, and do I really want that to be passed around?
It is not passed around, it's reference is passed around.

> I couldn't find any useful values on a ValidationResult other than errors.
Then it's fine I assumed custom formattes might be able to pass things into the object, but that's not the case.